docs/jep/11859-type-annotations.md

```diff
@@ -0,0 +1,192 @@
+- [google/pytype](https://github.com/google/pytype) is Google's static type checker, and projects which depend on JAX within Google frequently use this.
+- [microsoft/pyright](https://github.com/microsoft/pyright) is important as the static type checker used within VSCode for the Pylance completions mentioned previously.
+
+Full static type checking is the strictest of all the type annotation applications, because it will surface an error any time your type annotations are not precisely correct.
+On the one hand, this is nice because your static type analysis may catch faulty type annotations (for example, a case where a `DeviceArray` method is missing from the `jnp.ndarray` abstract class).
+
+On the other hand, this strictness can make the type checking process very brittle in packages that often rely on duck-typing rather than strict type-safe APIs.
+You'll currently find code comments like `#type: ignore` (for mypy) or `#pytype: disable` (for pytype) peppered throughout the JAX codebase in several hundred places.
+These typically represent cases where typing problems have arisen; they may be inaccuracies in JAX type annotations, or inaccuracies in the static type checker's ability to correctly follow the control flow in the code.
+On occasion, they are due to real & subtle bugs in the behavior of pytype or mypy.
+In rare cases, they may be due to the fact that JAX uses Python patterns that are difficult or even impossible to express in terms of Python's static type annotation syntax.
+
+## Type annotation challenges for JAX
+
+JAX currently has type annotations that are a mixture of different styles, and aimed at all three levels of type annotation discussed above.
+Partly, this comes from the fact that JAX's source code poses a number of unique challenges for Python's type annotation system. We'll outline them here.
+
+### Challenge 1: pytype, mypy and developer friction
+
+One challenge JAX currently faces is that package development must satisfy the constraints of two different static type checking systems, `pytype` (used by internal CI and internal Google projects) and `mypy` (used by external CI and external dependencies).
+Although the two type checkers have broad overlap in their behavior, each presents its own unique corner cases, as evidenced by the numerous `#type: ignore` and `#pytype: disable` statements throughout the JAX codebase.
+
+This creates friction in development: internal contributors may iterate until tests pass, only to find that on export their pytype-approved code falls afoul of mypy.
+For external contributors, it's often the opposite: a recent example is {jax-issue}`#9596` which had to be rolled-back after it failed internal Google continuous checks.
+Each time we move a type annotation from Level 1 (`Any` everywhere) to Level 2 or 3 (stricter annotations), it introduces more potential for such frustrating developer experiences.
+
+### Challenge 2: array duck-typing
```

Comment:
Do you think it might be possible to reduce the degree polymorphism at least in JAX internals? Public APIs could continue to use these highly polymorphic types, but the internals would only accept `DeviceArray`s.

Reply:
That's mostly the case already.